services/provider/server/consumer.go

```diff
@@ -0,0 +1,157 @@
+package server
+
+import (
+	"context"
+	"fmt"
+
+	ocsv1alpha1 "github.com/red-hat-storage/ocs-operator/api/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type consumerCache struct {
+	name   string
+	id     string
+	ticket string
+}
+
+type ocsConsumer struct {
+	client client.Client
+	cache  map[string]consumerCache
+}
+
+func NewOCSConsumer(ctx context.Context, client client.Client) (*ocsConsumer, error) {
+	cache, err := initConsumerCache(ctx, client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize consumer cache")
+	}
+	return &ocsConsumer{
+		client: client,
+		cache:  cache,
+	}, nil
+}
+
+func initConsumerCache(ctx context.Context, client client.Client) (map[string]consumerCache, error) {
+	consumers := &ocsv1alpha1.StorageConsumerList{}
+	err := client.List(ctx, consumers)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list storage consumers. %v", err)
+	}
+
+	cache := map[string]consumerCache{}
+	for _, consumer := range consumers.Items {
+		uid := string(consumer.UID)
+		ticket, ok := consumer.GetAnnotations()[TicketAnnotation]
+		if !ok {
+			continue
+		}
+		if _, ok := cache[uid]; !ok {
+			cache[uid] = consumerCache{consumer.Name, uid, ticket}
+		}
+	}
+
+	return cache, nil
+}
+
+// Create creates a new storageConsumer resource, updates the consumer cache and returns the storageConsumer UUID
+func (c *ocsConsumer) Create(ctx context.Context, consumerObj *ocsv1alpha1.StorageConsumer) (string, error) {
+	ticket := consumerObj.GetAnnotations()[TicketAnnotation]
+	if ticket == "" {
+		return "", fmt.Errorf("can't create storageConsumer resource without onboarding ticket annotation")
+	}
+
+	err := c.client.Create(ctx, consumerObj)
+	if err != nil {
+		if errors.IsAlreadyExists(err) {
+			return "", err
+		}
+		return "", fmt.Errorf("failed to create storageConsumer resource %q. %v", consumerObj.Name, err)
+	}
+
+	existingConsumer := &ocsv1alpha1.StorageConsumer{}
+	err = c.client.Get(ctx, types.NamespacedName{Name: consumerObj.Name, Namespace: consumerObj.Namespace}, existingConsumer)
+	if err != nil {
+		return "", fmt.Errorf("failed to get UUID from the storageConsumer resource %q. %v", consumerObj.Name, err)
+	}
```

Comment:
Ok, I see, you need this to get the UID. Are we sure that after calling create/update the resource is not already getting updated? I am asking because I remember that from a k8s API standpoint (not operators sdk standpoint) both create and update return the updated CR